src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter.PatternLocalRewriter.cs

```diff
@@ -516,8 +516,8 @@ private BoundExpression MakeSpanStringTest(BoundExpression input, ConstantValue
                     .Any(static node => node is BoundWhenDecisionDagNode { WhenExpression: { ConstantValue: null } });
 
                 var inputDagTemp = BoundDagTemp.ForOriginalInput(loweredInput);
-                if ((loweredInput.Kind == BoundKind.Local || loweredInput.Kind == BoundKind.Parameter)
-                    && loweredInput.GetRefKind() == RefKind.None &&
+                if (loweredInput.Kind is BoundKind.Local or BoundKind.Parameter &&
+                    loweredInput.GetRefKind() is RefKind.None or RefKind.In &&
```

Comment:
> or RefKind.In [](http://example.com/codeflow?start=61&length=14)

I am not sure why does this matter. The `in` modifier on the parameter doesn't provide a guarantee that the value doesn't change. It prevents this particular function to make the change.

Reply:
If, for example, we could detect that there is the only check for this local in the entire Dag, then we could avoid doing copy even for `ref` locals/parameters.
Alternatively, we could say that we do not care about changes made by other threads, and, given that we already assume that pattern matching is not side-effecting, we could careless about ref-ness and provide "good" code gen even for `ref` case.
Bottom line, I am not following why 'in' is somehow special here.

Reply:
Could you give a concrete example where doing the same optimization for `ref` is going to give an incorrect result?

Reply:
So if we take the following example, readonly safety is preserved in the current implementation, ie the output is "12"
```csharp
M(default(S));

void M(in S s)
{
    if (s.Value is 1) { Console.Write(1); }
    if (s.Value is 1) { Console.Write(2); }
}

struct S
{
    int item;
    public int Value => ++item;
}
```

If we do the same with a ref parameter, readonly safety is violated and the output is "1"
```csharp
var s = default(S);
M(ref s);

void M(ref S s)
{
    if (s.Value is 1) { Console.Write(1); }
    if (s.Value is 1) { Console.Write(2); }
}

struct S
{
    int item;
    public int Value => ++item;
}
```
(I have tested the scenarios locally)


Now these examples technically do violate the assumption of idempotency (no side-effecting), so we could optimize the `ref` case anyway, as you commented above. I will leave this decision up to the roslyn team.

Reply:
I am not sure if this is good example. We already observe this difference in behavior today, even without an optimization performed in this PR. And as far as I can see, the 'ref' parameter is not getting copied today.
```
    static void M_Ref(ref S s)
    {
        if (s.Value is 1) { Console.WriteLine(1); }
    }
```

The corresponding IL is:
```
    .method private hidebysig static 
        void M_Ref (
            valuetype S& s
        ) cil managed 
    {
        .maxstack 2
        .locals init (
            [0] bool
        )

        IL_0000: nop
        IL_0001: ldarg.0
        IL_0002: call instance int32 S::get_Value()
        IL_0007: ldc.i4.1
        IL_0008: ceq
        IL_000a: stloc.0
        IL_000b: ldloc.0
        IL_000c: brfalse.s IL_0017
        IL_000e: nop
        IL_000f: ldc.i4.1
        IL_0010: call void [System.Console]System.Console::WriteLine(int32)
        IL_0015: nop
        IL_0016: nop
        IL_0017: ret
    } // end of method Program::M_Ref
```

Reply:
I do not expect optimization to have any effect on this scenario.

Reply:
@ArcadeMode

> I will leave this decision up to the roslyn team.

 We discussed this offline and we fill comfortable with removing the ref kind check here completely. Let's go with that for now.

Unfortunately it looks like this change is not sufficient for improving code generated for this scenario:
```
class C
{
    static void PoorCodegen1(S s)
    {
        if (s.F is null) 
            System.Console.WriteLine();
    }
}

struct S
{
    public int? F;
}
```

It feels to me that we would want to improve this and the like scenarios as well. Could you investigate?